// @cloudflare/ai-chat/dist/types.js
//#region src/types.ts
/**
* Enum for message types to improve type safety and maintainability
*/
let MessageType = /* @__PURE__ */ function(MessageType) {
	MessageType["CF_AGENT_CHAT_MESSAGES"] = "cf_agent_chat_messages";
	MessageType["CF_AGENT_USE_CHAT_REQUEST"] = "cf_agent_use_chat_request";
	MessageType["CF_AGENT_USE_CHAT_RESPONSE"] = "cf_agent_use_chat_response";
	MessageType["CF_AGENT_CHAT_CLEAR"] = "cf_agent_chat_clear";
	MessageType["CF_AGENT_CHAT_REQUEST_CANCEL"] = "cf_agent_chat_request_cancel";
	/** Sent by server when client connects and there's an active stream to resume */
	MessageType["CF_AGENT_STREAM_RESUMING"] = "cf_agent_stream_resuming";
	/** Sent by client to acknowledge stream resuming notification and request chunks */
	MessageType["CF_AGENT_STREAM_RESUME_ACK"] = "cf_agent_stream_resume_ack";
	/** Client sends tool result to server (for client-side tools) */
	MessageType["CF_AGENT_TOOL_RESULT"] = "cf_agent_tool_result";
	/** Server notifies client that a message was updated (e.g., tool result applied) */
	MessageType["CF_AGENT_MESSAGE_UPDATED"] = "cf_agent_message_updated";
	/** Client sends tool approval response to server (for tools with needsApproval) */
	MessageType["CF_AGENT_TOOL_APPROVAL"] = "cf_agent_tool_approval";
	return MessageType;
}({});

//#endregion
export { MessageType };
